### PR TITLE
Windows fixes

### PR DESCRIPTION
**Pull request by tamiko**
This pull-request consists of the (hopefully) last bunch of workarounds that are necessary to compile the library with MSVC.

### CODE REVIEW

File: include/deal.II/base/numbers.h [hunk 1]
```diff
@@ -318,7 +318,7 @@ namespace numbers
   typename NumberTraits<number>::real_type
   NumberTraits<number>::abs (const number &x)
   {
-    return std::fabs(x);
+    return std::fabs((double)x);
```
Comment (tamiko) [inline include/deal.II/base/numbers.h]: This change has to be disussed. How to do this properly? MSVC seems to need an explicit cast here.

Comment (bangerth) [inline include/deal.II/base/numbers.h]: Yes, this doesn't work. How about just calling `std::abs()`? This should be overloaded for all floating point types.

File: include/deal.II/numerics/error_estimator.h [hunk 2]
```diff
@@ -337,7 +337,11 @@ class KellyErrorEstimator
                         const unsigned int       n_threads = numbers::invalid_unsigned_int,
                         const types::subdomain_id subdomain_id = numbers::invalid_subdomain_id,
                         const types::material_id       material_id = numbers::invalid_material_id,
+#ifdef DEAL_II_MSVC
+                        const Strategy strategy = cell_diameter_over_24);
+#else
                         const KellyErrorEstimator<dim,spacedim>::Strategy strategy = cell_diameter_over_24);
+#endif
```
Comment (bangerth) [inline include/deal.II/numerics/error_estimator.h]: Just always go with the `#if` branch. It is universally correct.

File: include/deal.II/numerics/error_estimator.h [hunk 3]
```diff
@@ -354,7 +358,11 @@ class KellyErrorEstimator
...
```
Comment (bangerth) [inline include/deal.II/numerics/error_estimator.h]: same here and below

File: source/numerics/error_estimator.cc [hunk 4]
```diff
@@ -954,7 +954,11 @@ estimate (const Mapping<dim, spacedim>      &mapping,
           const unsigned int       n_threads,
           const types::subdomain_id subdomain_id,
           const types::material_id       material_id,
+#ifdef DEAL_II_MSVC
+          const typename Strategy strategy)
+#else
```
Comment (bangerth) [inline source/numerics/error_estimator.cc]: No need for the `typename` here (I suppose, not knowing MSVC). It is in fact wrong in C++98.

**Comment by bangerth** [comment]
I think you should be able to get rid of most of the `#ifdef` statements and always go with the code that works with MSVC.

**Comment by tamiko** [comment]
@bangerth Even better - we can get rid of all #ifdef statements. I got sidetracked by the state superfluous ```typename```. Those produce compilation errors with gcc and I assumed that we also need the full namespace annotation in the header files...

**Comment by bangerth** [comment]
Nice cleanup. Thanks!

**Pull request merged**